Constructive Code Review/outline.txt

```diff
@@ -0,0 +1,205 @@
+    - Who's done code review before?
+        - Who's done it sync?
+            - We don't really do that, aside from maybe some pair
+              programming.
+        - Async?
+    - Pretty much all code at Mozilla gets reviewed before it gets
+      merged.
+        - It's a great way to…
+            - Improve the code
+                - I like to think I'm good at this, but people still
+                  find improvements to make in my stuff almost every
+                  time.
+            - Learn the codebase and the craft
+                - It informs other people what's going on in the
+                  codebase so you're not stuck with the sole
+                  responsibility of maintaining it.
+                - It's as much about teaching as fixing. It's a major
+                  way we bring junior people up the ladder.
+                - Tests may show the code works, but reviews ensure
+                  it's easy to understand.
+    - But open source runs on enthusiasm
+        - Even when you're getting paid to do it
+            - There's a lot of intrinsic motivation around here.
+    - “Crisitunity”
+        - We can drive the desire to contribute out of people, even
+          drive out the people themselves.
+        - Or we can magnify our power through teaching and
+          encouragement.
+        - I will show you how now, brown cow.
+        - For example…
+              …then continue on to goods/bads.
+- Good/bads
+      These will be a series of before/after PR screenshots with
+      discussion.
+    - Clarity
+          “This makes frob() buggy.” vs. “Now if we call frob(), there
+          will be a mem leak as we allocate the whatzit.”
+        - Sometimes I say "Hey, it would be cool if X, but it doesn't
+          block the merge."
+        - One step more is "We should X eventually. I think we should
+          either do it now or else open a bug about it for later."
+        - A big "That's all I see. With those changes, I think it's
+          ready to merge." lets them (and you!) know you're done and
+          ends with a note of encouragement.
+    - Softeners
+          When you have to point out something bad
+        - A spoonful of sugar
+        - The question mark
+              Dr. Evil pic: "My father…would make outrageous claims
+              like he invented the question mark."
+            - Cheap, easy sugar
+            - Because maybe you missed something, after all (saves
+              face!)
+            - “There's no point returning path results when there is
+              more than one term.” vs. “Can you remind me of some use
+              cases for returning path results when there is more than
+              one term?”
+            - “If I do this, it crashes. Do you get that as well?”
+            - Activates the Socratic method: great for teaching because
+              it makes people think along with you, not just get
+              battered with assertions
```

Comment:
This grouping is a good one because there are cases when the coder and the reviewer are both wrong. Posing questions and working out answers is a methodology that discovers that case a lot faster and without unnecessary emotional roughness.